// @solendprotocol/token2022-wrapper-sdk/dist/instructions/initializeWrapperToken.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createInitializeWrapperTokenInstruction = void 0;
const web3 = __importStar(require("@solana/web3.js"));
const types_1 = require("../types");
const constants_1 = require("../constants");
const utils_1 = require("../utils");
const BufferLayout = require("buffer-layout");
const createInitializeWrapperTokenInstruction = (payer, token2022Mint) => {
    const dataLayout = BufferLayout.struct([BufferLayout.u8("instruction")]);
    const data = Buffer.alloc(dataLayout.span);
    dataLayout.encode({
        instruction: types_1.TokenWrapperInstruction.InitializeWrapperToken,
    }, data);
    const wrapperTokenMint = (0, utils_1.getWrapperTokenMint)(token2022Mint);
    const reserveAuthority = (0, utils_1.getReserveAuthority)(token2022Mint);
    const reserveAuthorityTokenAccount = (0, utils_1.getReserveAuthorityTokenAccount)(token2022Mint);
    const keys = [
        {
            pubkey: payer,
            isSigner: true,
            isWritable: true,
        },
        {
            pubkey: token2022Mint,
            isSigner: false,
            isWritable: false,
        },
        {
            pubkey: wrapperTokenMint,
            isSigner: false,
            isWritable: true,
        },
        {
            pubkey: reserveAuthority,
            isSigner: false,
            isWritable: true,
        },
        {
            pubkey: reserveAuthorityTokenAccount,
            isSigner: false,
            isWritable: true,
        },
        {
            pubkey: constants_1.TOKEN_PROGRAM_ID,
            isSigner: false,
            isWritable: false,
        },
        {
            pubkey: constants_1.TOKEN_2022_PROGRAM_ID,
            isSigner: false,
            isWritable: false,
        },
        {
            pubkey: constants_1.SYSTEM_PROGRAM_ID,
            isSigner: false,
            isWritable: false,
        },
        {
            pubkey: constants_1.RENT_SYSVAR,
            isSigner: false,
            isWritable: false,
        },
    ];
    return new web3.TransactionInstruction({
        keys,
        programId: constants_1.PROGRAM_ID,
        data,
    });
};
exports.createInitializeWrapperTokenInstruction = createInitializeWrapperTokenInstruction;
